Fix: locale redirect on pricing page

andrii-lysenko opened the pull request:
#### Changes proposed in this Pull Request

Currently, the checkout page on Calypso doesn't take language into account. It means when redirected from cloud.jetpack.com and jetpack.com to checkout, not logged in, language will be English by default. We are adding to two URLs: `/checkout/jetpack/:productSlug` and `/checkout/jetpack/:siteSlug/:productSlug` lang param in the end to handle language.
In this PR we are also passing it from cloud.jetpack.com side.

Note: jetpack.com -> wordpress.com/checkout... will be handled separately.

#### Testing instructions
To test locally:
- Checkout the branch locally.
- `yarn install`
- `yarn start-jetpack-cloud` 
- `PORT=443 PROTOCOL=https yarn start`
- Add next line to hosts file (or using gasmask) `127.0.0.1 wordpress.com` and 'flush dns'

- navigate to http://jetpack.cloud.localhost:3000/pricing , then try to purchase a product. You should be redirected correctly to `wordpress.com/checkout...` with default language.
- navigate to  http://jetpack.cloud.localhost:3000/fr/pricing or change the language manually to not default. You should be redirected to a proper checkout page with the language selected previously.
- These steps should also work with site selected when purchase on `checkout/jetpack`. On cloud jetpack when `canDoUnlinkedCheckout` is set to true.
- Need also to check that all links to checkout from cloud.jetpack.com and jetpack.com to wordpress.com/checkout are working properly.

Related to #
1164141197617539-as-1201508699348275/f

Discussion:
p9oQ9f-10K-p2

andrii-lysenko committed:
update: add localized jetpack checkout URLs

andrii-lysenko committed:
update: pass lang param to while redirecting to checkout

andrii-lysenko committed:
Merge remote-tracking branch 'origin' into fix/locale-redirect-on-pricing-page

andrii-lysenko committed:
update: switch to query args

andrii-lysenko committed:
Merge remote-tracking branch 'origin/trunk' into fix/locale-redirect-on-pricing-page

andrii-lysenko committed:
update: back to query params

andrii-lysenko committed:
Merge remote-tracking branch 'origin' into fix/locale-redirect-on-pricing-page

andrii-lysenko committed:
fix: URLs

andrii-lysenko committed:
update: fix broken url

nateweller commented on client/my-sites/plans/jetpack-plans/controller.tsx:
```diff
@@ -87,6 +89,7 @@ export function jetpackFreeWelcome( context: PageJS.Context, next: () => void ):
 export const jetpackStoragePricing = ( context: PageJS.Context, next: () => void ) => {
 	const { site, duration } = getParamsFromContext( context );
 	const urlQueryArgs: QueryArgs = context.query;
+	const { locale } = context.params;
```
`locale` is `undefined` here when I try to load a URL such as: 
http://jetpack.cloud.localhost:3001/fr/pricing/storage/odd-shrew.jurassic.ninja

I think this is because this URL redirects to:
http://jetpack.cloud.localhost:3001/pricing/storage/odd-shrew.jurassic.ninja

Somehow it keeps the french language on the page, but loses the `context.params` value because the language is removed from the URL. Due to this, the buttons pointing to the checkout page are not including the `/fr` locale.

One solution would be to implement the `useLocale()` hook from `@automattic/i18n-utils` inside of the `<StoragePricing />` component instead of passing it in from `context.params`. 

It might also make sense to update this page to keep the language in the URL, like we did with the main `/pricing` page, and then the `context.params` will have the value we would expect to have 👍

jsnajdr replied:
Instead of getting the locale from `context.params` which only looks at the current URL, get it from the `i18n` object itself, which returns the current locale no matter how it was set -- from URL or anything else:
```js
function SelectorPage() {
  const { locale } = useTranslate();
  buildCheckoutUrl( locale );
}
```
or with a HOC:
```js
function SelectorPage( { locale /* injected by `localize` HOC */ } ) {
  buildCheckoutUrl( locale );
}
export default localize( SelectorPage );
```
That should fix the issues that @nateweller reports.

andrii-lysenko replied:
@jsnajdr Thank you for the comment. So currently,`buildCheckoutUrl ` isn't used directly, but is wrapped in a closure which generated by `getPurchaseURLCallback`. The closure is passed through several elements and I'm not totally sure how to use the selector in this case. 
I could possibly use `useLocale()` inside `ProductCard` element (where actually URLs are generated and used) which will eliminate getting the locale param from only URL, though, I have a small concern that it will actually retrieve current locale, which means that redirecting to checkout page will **always** include language param (currently, it's mostly for the users which are not logged in and selected non default language on cloud.jetpack.com). What do you think?
cc @nateweller

jsnajdr replied:
> I have a small concern that it will actually retrieve current locale, which means that redirecting to checkout page will always include language param 

Yes, this is indeed what would happen: the page is always in _some_ language, and the checkout URL would include that locale slug. If the desired behavior is "include language param in the link only if the current page includes a language param, too", then reading from `context.params` is right after all :+1:

andrii-lysenko replied:
Thank you and thanks @nateweller for finding the issue. Will try to include all other fixes and fix current behaviour.

jsnajdr commented on client/my-sites/checkout/index.js:
```diff
@@ -32,6 +38,15 @@ export default function () {
 	);
 	page( '/checkout/jetpack/:productSlug', noSite, checkoutSiteless, makeLayout, clientRender );
+	page(
+		`/checkout/jetpack/:productSlug/${ getLanguageRouteParam() }`,
```
You don't need to create a second `/checkout/jetpack/:productSlug` route handler. Just add the lang suffix to the original one. By default it's optional. And `setLocaleMiddleware` will act as noop if the lang param is not present.

jsnajdr commented on client/my-sites/checkout/index.js:
```diff
@@ -83,6 +98,15 @@ export default function () {
 	page( '/checkout/jetpack/:siteSlug/:productSlug', checkout, makeLayout, clientRender );
+
+	page(
+		`/checkout/jetpack/:siteSlug/:productSlug/${ getLanguageRouteParam() }`,
```
Same here: let's add the locale handling to the original route instead of creating a duplicate one.

mattgawarecki reviewed (commented):
I'll wait to thoroughly test until after conflicts are resolved, but code-wise this is looking great to me so far. I highly recommend making the changes @jsnajdr suggested, as well. I appreciate your use of unit tests, @andrii-lysenko -- very nice!

andrii-lysenko committed:
update: use locale param when site is selected

andrii-lysenko committed:
update: fix linter

jsnajdr commented on client/jetpack-cloud/controller.ts:
```diff
@@ -206,7 +207,11 @@ export async function cloudSiteSelection(
-		if ( context.path.startsWith( '/pricing' ) ) {
+		const languages = getLanguageSlugs().join( '|' );
+		// Regex defines url starts with /:locale/pricing or /pricing
+		const pricingMatchingPath = new RegExp( `(^/(${ languages })/pricing)|^/pricing` );
```
The regexp could be more compact, like `^(/(${languages}))?/pricing`.

manzoorwanijk replied:
Yes, I was going to suggest the same, then thought that it might be more readable 😄

manzoorwanijk commented on client/jetpack-cloud/controller.ts:
```diff
@@ -206,7 +207,11 @@ export async function cloudSiteSelection(
-		if ( context.path.startsWith( '/pricing' ) ) {
+		const languages = getLanguageSlugs().join( '|' );
+		// Regex defines url starts with /:locale/pricing or /pricing
+		const pricingMatchingPath = new RegExp( `(^/(${ languages })/pricing)|^/pricing` );
```
A bit nit-picky, but since we don't use the capturing groups, we can optimize it a little by using [non-capturing groups](https://www.regular-expressions.info/brackets.html).

```suggestion
		const pricingMatchingPath = new RegExp( `(?:^/(?:${ languages })/pricing)|^/pricing` );
```

jsnajdr commented on client/jetpack-cloud/sections/pricing/controller.tsx:
```diff
@@ -8,18 +8,11 @@ import JetpackComMasterbar from './jpcom-masterbar';
 	const path = context.path;
 
 	if ( locale ) {
 		context.store.dispatch( setLocale( locale ) );
```
Instead of this `setLocale` call in the `jetpackPricingContext` handler, the route handler should use the specialized `setLocaleMiddleware` handler.

andrii-lysenko committed:
update: PR refactor and fix

andrii-lysenko committed:
Merge branch 'trunk' into fix/locale-redirect-on-pricing-page

andrii-lysenko committed:
update: fix pr

andrii-lysenko commented:
@jsnajdr @nateweller @manzoorwanijk  thank you guys for your comments. I've fixed the issue and refactored the suggested parts. Could you please take a look if I haven't missed anything? Thanks!

jsnajdr commented on client/lib/route/path.ts:
```diff
@@ -41,8 +41,8 @@ export function getSiteFragment( path: URLString ): SiteSlug | SiteId | false {
-	// Check last and second-to-last piece for site slug
-	for ( let i = 2; i > 0; i-- ) {
+	// Check last, second-to-last and third-to-last (in case of locale param)  piece for site slug
+	for ( let i = 3; i > 0; i-- ) {
```
This patch is needed to make the
```
/checkout/jetpack/:siteSlug/:productSlug/:lang
```
route work correctly? You could avoid changing this function if the site param was called `:site` instead. The code that determines the site fragment looks like:
```js
siteFragment = context.params.site || getSiteFragment( context.path );
```
It first looks at the `:site` param, and only if it's not present, it does the unreliable heuristic in `getSiteFragment`.

Better to call the param `:site` and avoid the heuristic if possible.

andrii-lysenko replied:
So it's actually this one: `/checkout/:product/:domainOrProduct/:lang` and in this case it's product. There are quite a lot of middleware functions (`redirectLoggedOut, siteSelection, redirectJetpackLegacyPlans, checkout, makeLayout, clientRender`) so I'm a bit worried I could broke something deeper inside? Or should we still do the refactoring here?

jsnajdr replied:
Here the site slug is in `:domainOrProduct`, never in `:product`, is it? So looking at the second-to-last piece is enough, we don't need to look at the third.

What does the `/checkout/:product/:domainOrProduct` route display anyway? The `checkout` controller never reads the `:product` or `:domainOrProduct` params, it reads a set of different ones, none of which is present in the route.

andrii-lysenko replied:
I'm very new to Calypso and maybe lacking some context. So, on `jetpack.cloud` side in `get-purchase-url-callback` I see that URL is build like: `/checkout/${ siteSlug }/${ productsString }` -> `/checkout/:product/:domainOrProduct` so `product = siteSlug`, and  `domainOrProduct = productsString`. It renders a checkout page for given site and product. So, in this case, I would probably need to rename 'product' into 'site', and then refactor all middleware functions like `getDomainOrProductFromContext` and make sure if any other registered pages rely on it.

jsnajdr replied:
I think I undestand the checkout routes a bit better now. The routes affected by this PR, the ones that get a new `:lang` suffix, are:
- `/checkout/jetpack/:siteSlug/:productSlug`: here we can rename the params to `:site` and `:product` and change the places that read from `context.params`. There are just 2-3 of them. The benefits are that `siteSelection` will pick up the `:site` param automatically, without `getSiteFragment`, and that the `:product` param name is now more consistent with other checkout routes
- `/checkout/:domainOrProduct`: this is the only one where params are ambiguous: the route can be either `/checkout/example.blog`, showing checkout for the `example.blog` site, without adding any products, or `/checkout/plan_name`, selecting no site and adding `plan_name` to cart. Here the param can be renamed to `:site` and then we need to also change `getDomainOrProductFromContext` to read from `params.site` instead of `params.domainOrProduct`.
- `/checkout/:product/:domainOrProduct`: here the param naming seems completely wrong, but accidentally it all works. `:product` is always a site, and `siteSelection` will read it with `getSiteFragment`, no matter what the param name is. And the product slug will be read from `domainOrProduct` with `getDomainOrProductFromContext`. The route should be renamed to `/checkout/:site/:product`. Then `siteSelection` will read the site slug from `:site`, even if it's the third last (after `:lang` is appended), and `getDomainOrProductFromContext` will read product slug from `:product`.

What do you think? The proposed changes would clean up the param naming and would make the routing code more comprehensible.

andrii-lysenko replied:
Thanks for the clarification. Will try to refactor this part.

andrii-lysenko replied:
@jsnajdr - updated. Could you please take a look if I haven't missed anything? Thanks!

jsnajdr reviewed (commented):
This looks good and is very close to being shippable :+1: I'm only suggesting to avoid patching the `getSiteFragment` function.

andrii-lysenko committed:
update: refactor checkout param namings

jsnajdr reviewed (approved):
Thanks for implementing the route param changes, they look good :+1: The last thing before merging would be to remove the now-redundant changes in `getSiteFragment`.

andrii-lysenko committed:
update: remove unecessary logic

elliottprogrammer reviewed (approved):
Tested cloud.jetpack.com/:lang/pricing -> /checkout/jetpack/:product while fully logged-out and also tested various Cloud(with different locale) -> /checkout flows while logged-in also. This all seems to be working good (Checkout page displaying in selected locale) in all my testing. I don't see any issues here.
LGTM! 👍

mattgawarecki commented:
In functional testing (using Japanese as my chosen language), everything appears to work for me. I did notice a couple slight changes in copy might've crept in while this PR was developed, so I recommend a rebase before deploying. Even though GitHub says there's no conflict with `origin/trunk`, I ran into one almost immediately when rebasing locally.

andrii-lysenko committed:
Merge remote-tracking branch 'origin' into fix/locale-redirect-on-pricing-page

andrii-lysenko committed:
fix imports

andrii-lysenko committed:
Merge remote-tracking branch 'origin' into fix/locale-redirect-on-pricing-page

andrii-lysenko committed:
fix imports

sirbrillig commented on client/my-sites/checkout/index.js:
```diff
@@ -199,7 +218,8 @@ export default function () {
 	);
 
 	page(
-		'/checkout/:domainOrProduct',
```
With this removal, I think that the route `/checkout/:productSlug` will no longer work correctly. Preparing a revert in https://github.com/Automattic/wp-calypso/pull/62225
